presto-main/src/main/java/io/prestosql/operator/scalar/JoniRegexpFunctions.java

```diff
@@ -326,4 +327,114 @@ private static void validateGroup(long group, Region region)
+                                      @SqlType(StandardTypes.INTEGER) long occurrence)
+    {
+        // start position cannot be smaller than 1
+        if (start < 1) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "start position cannot be smaller than 1");
+        }
+        // occurrence cannot be smaller than 1
+        if (occurrence < 1) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "occurrence cannot be smaller than 1");
+        }
+        // returns -1 if start is greater than the length of source
+        if (start > SliceUtf8.countCodePoints(source)) {
+            return -1;
+        }
+
+        Matcher matcher = pattern.matcher(source.getBytes());
+        long count = 0;
+        // convert char position to byte position
+        // subtract 1 because codePointCount starts from zero
+        int nextStart = SliceUtf8.offsetOfCodePoint(source, (int) start - 1);
+        while (true) {
+            int offset = matcher.search(nextStart, source.length(), Option.DEFAULT);
+            // Check whether offset is negative, offset is -1 if no pattern was found or -2 if process was interrupted
+            if (offset < 0) {
+                return -1;
+            }
+
+            if (++count == occurrence) {
+                // Plus 1 because position returned start from 1
+                return SliceUtf8.countCodePoints(source, 0, matcher.getBegin()) + 1;
+            }
+
+            if (matcher.getEnd() == matcher.getBegin()) {
+                if (matcher.getBegin() < source.length()) {
+                    nextStart = matcher.getEnd() + lengthOfCodePointFromStartByte(source.getByte(matcher.getBegin()));
+                }
+                else {
+                    // last match is empty and we matched end of source, move past the source length to terminate the loop
+                    nextStart = matcher.getEnd() + 1;
+                }
+            }
+            else {
+                nextStart = matcher.getEnd();
+            }
+        }
+    }
+
+    @ScalarFunction
+    @Description("returns the number of times that a pattern occurs in a string")
+    @LiteralParameters("x")
+    @SqlType(StandardTypes.BIGINT)
+    public static long regexpCount(@SqlType("varchar(x)") Slice source, @SqlType(JoniRegexpType.NAME) JoniRegexp pattern)
+    {
+        Matcher matcher = pattern.matcher(source.getBytes());
+
+        int count = 0;
+        // Start from zero, implies the first byte
+        int nextStart = 0;
+        while (true) {
+            // mather.search returns `source.length` if `nextStart` equals `source.length - 1`.
+            // It should return -1 if `nextStart` is greater than `source.length - 1`.
+            int offset = matcher.search(nextStart, source.length(), Option.DEFAULT);
+            if (offset < 0) {
+                break;
+            }
+
+            if (matcher.getEnd() == matcher.getBegin()) {
```

Comment:
I think we could extract a method out of it, e.g `int getNextStart(Slice source, Matcher matcher)`

Reply:
Great! This method have been added.